token-metadata/program/src/processor/metadata/update.rs

```diff
@@ -95,46 +89,77 @@ fn update_v1(program_id: &Pubkey, ctx: Context<Update>, args: UpdateArgs) -> Pro
+            .ok_or(MetadataError::MissingAuthorizationRulesProgram)?;
+
+        if authorization_rules_program.key != &mpl_token_auth_rules::ID {
+            return Err(ProgramError::IncorrectProgramId);
         }
     }
 
     // Validate relationships
 
+    // Token
+    let (token_pubkey, token) = if let Some(token_info) = ctx.accounts.token_info {
+        let token = Account::unpack(&token_info.try_borrow_data()?)?;
+
+        // Token mint must match mint account key.
+        if token.mint != *ctx.accounts.mint_info.key {
+            return Err(MetadataError::MintMismatch.into());
+        }
+
+        (Some(token_info.key), Some(token))
+    } else {
+        (None, None)
+    };
+
+    // Metadata
     let mut metadata = Metadata::from_account_info(ctx.accounts.metadata_info)?;
-    // Mint must match metadata mint
+    // Metadata mint must match mint account key.
     if metadata.mint != *ctx.accounts.mint_info.key {
         return Err(MetadataError::MintMismatch.into());
     }
 
-    let token_standard = if let Some(token_standard) = metadata.token_standard {
-        token_standard
-    } else {
-        check_token_standard(ctx.accounts.mint_info, ctx.accounts.edition_info)?
-    };
+    // Edition
+    if let Some(edition) = ctx.accounts.edition_info {
+        // checks that we got the correct edition account
+        assert_derivation(
+            program_id,
+            edition,
+            &[
+                PREFIX.as_bytes(),
+                program_id.as_ref(),
+                ctx.accounts.mint_info.key.as_ref(),
+                EDITION.as_bytes(),
+            ],
+        )?;
+    }
 
-    let (token_pubkey, token) = if let Some(token_info) = ctx.accounts.token_info {
-        (
-            Some(token_info.key),
-            Some(Account::unpack(&token_info.try_borrow_data()?)?),
-        )
-    } else {
-        (None, None)
-    };
+    // Check authority.
 
-    // there is a special case for collection-level delegates, where the
-    // validation should use the collection key as the mint parameter
-    let collection_mint = if let Some(Collection { key, .. }) = &metadata.collection {
-        Some(key)
-    } else {
-        None
+    // There is a special case for collection-level delegates, where the
+    // validation should use the collection key as the mint parameter.
+    let existing_collection_mint = metadata
+        .collection
+        .as_ref()
+        .map(|Collection { key, .. }| key);
+
+    // Check if caller passed in a collection and if so use that.  Note that
+    // `validate_update` checks that the authority has permission to pass in
+    // a new collection value.
+    let collection_mint = match &args {
+        UpdateArgs::V1 { collection, .. }
+        | UpdateArgs::AsUpdateAuthorityV2 { collection, .. }
+        | UpdateArgs::AsCollectionDelegateV2 { collection, .. }
+        | UpdateArgs::AsCollectionItemDelegateV2 { collection, .. } => match collection {
```

Comment:
Could this be collapsed?

Reply:
I don't think I can collapse it any further because its already piped into a single case.  My previous design used `InternalUpdateArgs` and used the From trait to destructure everything into one of those, so I had the superior code:
```
let internal_update_args = InternalUpdateArgs::from(args);
match internal_update_args.collection {
   ...
}
```
But unfortunately this added about 5k to the binary size.

I went back with @febo today and attempted to do it again without the `From` trait, implementing a partial use of `InternalUpdateArgs` to just avoid some of the duplicate destructuring, but even this used about 2k more bytes.  That way didn't change this particular collection destructuring but did help with another one and reduce the overall need for this type of destructure.  Here is that attempt for reference: https://github.com/metaplex-foundation/metaplex-program-library/pull/1079

Overall this raw destructuring and doing it multiple times, rather than DRY code with a new type, seems to result in the smallest binary size.